src/s3ql/http.py

```diff
@@ -0,0 +1,1558 @@
+                if self.trace_fh:
+                    self.trace_fh.write(buf2)
+                buf += buf2
+            else:
+                buf += rbuf.exhaust()
+
+        return buf
+
+    async def _co_read_id(self, len_):
+        '''Read up to *len* bytes of response body assuming identity encoding'''
+
+        log.debug('start (len=%d)', len_)
+        assert self._in_remaining is not None
+
+        if not self._in_remaining:
+            # Body retrieved completely, clean up
+            self._in_remaining = None
+            self._pending_requests.popleft()
+            return b''
+
+        rbuf = self._rbuf
+        if self._in_remaining is not READ_UNTIL_EOF:
+            len_ = min(len_, self._in_remaining)
+        log.debug('updated len_=%d', len_)
+
+        # If buffer is empty, reset so that we start filling from
+        # beginning. This check is already done by _try_fill_buffer(), but we
+        # have to do it here or we never enter the while loop if the buffer
+        # is empty but has no capacity (rbuf.b == rbuf.e == len(rbuf.d))
+        if rbuf.b == rbuf.e:
+            rbuf.b = 0
+            rbuf.e = 0
+
+        # Loop while we could return more data than we have buffered
+        # and buffer is not full
+        while len(rbuf) < len_ and rbuf.e < len(rbuf.d):
+            got_data = self._try_fill_buffer()
+            if got_data is None:
+                if rbuf:
+                    log.debug('nothing more to read')
+                    break
+                else:
+                    log.debug('buffer empty and nothing to read, yielding..')
+                    await PollNeeded(self._sock.fileno(), POLLIN)
+            elif got_data == 0:
+                if self._in_remaining is READ_UNTIL_EOF:
+                    log.debug('connection closed, %d bytes in buffer', len(rbuf))
+                    self._in_remaining = len(rbuf)
+                    break
+                else:
+                    raise ConnectionClosed('server closed connection')
+
+        len_ = min(len_, len(rbuf))
+        if self._in_remaining is not READ_UNTIL_EOF:
+            self._in_remaining -= len_
+
+        if len_ < len(rbuf):
+            buf = rbuf.d[rbuf.b : rbuf.b + len_]
+            rbuf.b += len_
+        else:
+            buf = rbuf.exhaust()
+
+        # When reading until EOF, it is possible that we read only the EOF. In
+        # this case we won't get called again, so we need to clean-up the
+        # request. This can not happen when we know the number of remaining
+        # bytes, because in this case either the check at the start of the
+        # function hits, or we can't read the remaining data and raise.
+        if len(buf) == 0:
+            assert self._in_remaining == 0
+            self._in_remaining = None
+            self._pending_requests.popleft()
+
+        log.debug('done (%d bytes)', len(buf))
+        return buf
+
+    async def _co_read_chunked(self, len_: int):
```

Comment:
If you're going to do type annotation, does it make sense to do it consistently?

Reply:
Ideally, yes. But currently I don't have enough spare time to annotate the entire codebase, so I'm just doing it when I modify something :-).